File: Examples/Comparison/consensus_support_newick_compare.py
import random
from io import StringIO
from time import perf_counter

from Bio import Phylo
from Bio.Align import MultipleSeqAlignment
from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio.Phylo import Consensus
from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor


def build_records(record_count=120, width=80):
    template = ("ACGT" * ((width + 3) // 4))[:width]
    bases = "ACGT"
    records = []
    for record_index in range(1, record_count + 1):
        chars = list(template)
        for position in range(1, width + 1):
            if (record_index + position) % 17 == 0:
                chars[position - 1] = "-"
            elif (record_index * position) % 19 == 0:
                chars[position - 1] = bases[(record_index + position - 2) % len(bases)]
        records.append(SeqRecord(Seq("".join(chars)), id=f"seq{record_index}", description="bootstrap_input"))
    return records


def time_call(func):
    start = perf_counter()
    result = func()
    return (perf_counter() - start) * 1000.0, result


def bootstrap_consensus(alignment):
    calculator = DistanceCalculator("identity")
    constructor = DistanceTreeConstructor(calculator)
    trees = list(Consensus.bootstrap_trees(alignment, 25, constructor))
    consensus = Consensus.majority_consensus(trees, cutoff=0.5)
    output = StringIO()
    Phylo.write([consensus], output, "newick")
    return output.getvalue()


def main():
    random.seed(2024)
    alignment = MultipleSeqAlignment(build_records())
    consensus_ms, newick = time_call(lambda: bootstrap_consensus(alignment))

    assert "(" in newick
    assert len(alignment) == 120

    print(f"python_consensus_support_newick_ms={consensus_ms:.4f}")
    print(f"python_consensus_support_newick_ok={str(True).lower()}")


if __name__ == "__main__":
    main()

File: Examples/Comparison/consensus_support_newick_compare.jl
pushfirst!(LOAD_PATH, abspath(joinpath(@__DIR__, "..", "..")))

using Random
using BioToolkit
include(joinpath(@__DIR__, "..", "..", "scripts", "benchmark_helpers.jl"))

function build_records(record_count::Integer=120, width::Integer=80)
    Random.seed!(2024)
    template = repeat("ACGT", cld(width, 4))[1:width]
    bases = ('A', 'C', 'G', 'T')
    records = SeqRecordLite[]

    for record_index in 1:record_count
        chars = collect(template)
        for position in 1:width
            if mod(record_index + position, 17) == 0
                chars[position] = '-'
            elseif mod(record_index * position, 19) == 0
                chars[position] = bases[mod(record_index + position - 2, length(bases)) + 1]
            end
        end
        push!(records, SeqRecordLite(String(chars); identifier="seq$(record_index)", description="bootstrap_input"))
    end

    return records
end

function main()
    alignment = MultipleSequenceAlignment(build_records())
    bootstrap_ms, consensus = repeat_elapsed_ms(() -> BioToolkit.bootstrap_consensus_tree(alignment; replicates=25, threshold=0.5, method=:hamming, constructor=:nj), 5)

    newick = BioToolkit.write_newick(consensus)
    @assert occursin("0.", newick) || occursin("1.0", newick)
    @assert length(BioToolkit.get_terminals(consensus)) == length(alignment.records)

    python_output = read(`conda run -n general --no-capture-output python $(joinpath(@__DIR__, "consensus_support_newick_compare.py"))`, String)
    python_consensus_ms = parse(Float64, match(r"python_consensus_support_newick_ms=([0-9.]+)", python_output).captures[1])
    python_consensus_ok = match(r"python_consensus_support_newick_ok=(true|false)", python_output).captures[1] == "true"

    @assert python_consensus_ok

    println("Consensus support Newick benchmark")
    println("  record_count=120")
    println("  julia_consensus_support_newick_ms=$(round(bootstrap_ms / 5; digits=4))")
    println("  python_consensus_support_newick_ms=$(round(python_consensus_ms; digits=4))")
    println("  julia_newick_length=$(length(newick))")
    print(python_output)
end

main()
